### PR TITLE
refactor(compiler): extract BindingParser

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

File: modules/@angular/compiler/src/template_parser/binding_parser.ts [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+const PROPERTY_PARTS_SEPARATOR = '.';
+const ATTRIBUTE_PREFIX = 'attr';
+const CLASS_PREFIX = 'class';
+const STYLE_PREFIX = 'style';
+
+const ANIMATE_PROP_PREFIX = 'animate-';
+
+/**
+ * Type of a parsed property
+ */
```
Comment (vicb) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: As much as I love inline docs, I'm not sur this one (and the one below) is really helpful.

Would be more useful to explain each of the values here for ex.

File: modules/@angular/compiler/src/template_parser/binding_parser.ts [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+  errors: ParseError[] = [];
+
+  constructor(
+      private _exprParser: Parser, private _interpolationConfig: InterpolationConfig,
+      protected _schemaRegistry: ElementSchemaRegistry, protected _schemas: SchemaMetadata[],
+      pipes: CompilePipeMetadata[]) {
+    pipes.forEach(pipe => this.pipesByName.set(pipe.name, pipe));
+  }
+
+  createDirectiveHostPropertyAsts(
```
Comment (vicb) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: nit: `create...` but returns nothing.... `accumualte...` ?

File: modules/@angular/compiler/src/template_parser/binding_parser.ts [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+              `Value of the host property binding "${propName}" needs to be a string representing an expression but got "${expression}" (${typeof expression})`,
+              sourceSpan);
+        }
+      });
+      boundProps.forEach(
+          (prop) => { targetPropertyAsts.push(this.createElementPropertyAst(elementName, prop)); });
+    }
+  }
+
+  createDirectiveHostEventAsts(
```
Comment (vicb) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: same comment from "create..."

File: modules/@angular/compiler/src/template_parser/binding_parser.ts [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+      });
+    }
+  }
+
+  parseInterpolation(value: string, sourceSpan: ParseSourceSpan): ASTWithSource {
+    const sourceInfo = sourceSpan.start.toString();
+
+    try {
+      const ast = this._exprParser.parseInterpolation(value, sourceInfo, this._interpolationConfig);
+      if (ast) this._reportExpressionParserErrors(ast.errors, sourceSpan);
```
Comment (vicb) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: `if (ast.errors)` ???

don't have tests ?

File: modules/@angular/compiler/src/template_parser/binding_parser.ts [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+      return [];
+    }
+  }
+
+  parseLiteralAttr(
+      name: string, value: string, sourceSpan: ParseSourceSpan, targetMatchableAttrs: string[][],
+      targetProps: BoundProperty[]) {
+    if (_isAnimationLabel(name)) {
+      name = name.substring(1);
+      if (isPresent(value) && value.length > 0) {
```
Comment (vicb) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: if (value)

File: modules/@angular/compiler/src/template_parser/binding_parser.ts [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+      if (ast) this._reportExpressionParserErrors(ast.errors, sourceSpan);
+      this._checkPipes(ast, sourceSpan);
+      return ast;
+    } catch (e) {
+      this.reportError(`${e}`, sourceSpan);
+      return this._exprParser.wrapLiteralPrimitive('ERROR', sourceInfo);
+    }
+  }
+
+  createElementPropertyAst(elementName: string, boundProp: BoundProperty): BoundElementPropertyAst {
```
Comment (vicb) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: nit: I prefer to have public methods first (API) and then protected and private.

File: modules/@angular/compiler/src/template_parser/template_parser.ts [hunk 7]
```diff
@@ -195,135 +196,23 @@ export class TemplateParser {
   }
 }
 
-class TemplateParseVisitor implements html.Visitor {
+class TemplateParseVisitor extends BindingParser implements html.Visitor {
```
Comment (vicb) [inline modules/@angular/compiler/src/template_parser/template_parser.ts]: would it be possible to use composition instead ?

File: modules/@angular/compiler/src/template_parser/binding_parser.ts (hunk 4 above)
Comment (vicb) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: seems like a cut & paste, so it must actually work

File: modules/@angular/core/src/linker/view_utils.ts [hunk 8]
```diff
@@ -377,3 +377,153 @@ const CAMEL_CASE_REGEXP = /([A-Z])/g;
...
+        return undefined;
+    }
+  }
+}
+
+export class FastArrayDynamic<T> implements FastArray<T> {
+  private _values: any[];
+  // Note: We still take the length argument so this class can be created
+  // in the same ways as the other classes!
+  constructor(public length: number, ...values: any[]) { this._values = values; }
```
Comment (vicb) [inline modules/@angular/core/src/linker/view_utils.ts]: can you use `private ..._values`  ?

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts [hunk 9]
```diff
@@ -45,74 +46,83 @@ export class DirectiveWrapperCompiler {
...
+  constructor(public compilerConfig: CompilerConfig, public dirMeta: CompileDirectiveMetadata) {
+    const dirLifecycleHooks = dirMeta.type.lifecycleHooks;
+    this.genChanges = dirLifecycleHooks.indexOf(LifecycleHooks.OnChanges) !== -1 ||
+        this.compilerConfig.logBindingUpdate;
+    this.ngOnChanges = dirLifecycleHooks.indexOf(LifecycleHooks.OnChanges) !== -1;
+    this.ngOnInit = dirLifecycleHooks.indexOf(LifecycleHooks.OnInit) !== -1;
+    this.ngDoCheck = dirLifecycleHooks.indexOf(LifecycleHooks.DoCheck) !== -1;
+  }
+
+  build(): o.ClassStmt {
```
Comment (vicb) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: build -> toStatements ?

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts [hunk 10]
```diff
@@ -45,74 +46,83 @@ export class DirectiveWrapperCompiler {
...
+  methods: o.ClassMethod[] = [];
+  ctorStmts: o.Statement[] = [];
+
+  genChanges: boolean;
+  ngOnChanges: boolean;
+  ngOnInit: boolean;
+  ngDoCheck: boolean;
+
+  constructor(public compilerConfig: CompilerConfig, public dirMeta: CompileDirectiveMetadata) {
+    const dirLifecycleHooks = dirMeta.type.lifecycleHooks;
```
Comment (vicb) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: don't we just do some extra processing to generate this array and have to indexOf here again ?

couldn't we just store a `{hook: boolean, ...}` in this prop ? (out of scope for this PR)

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts [hunk 11]
```diff
@@ -45,74 +46,83 @@ export class DirectiveWrapperCompiler {
...
-        ctorStmts);
 
-    const wrapperClassName = DirectiveWrapperCompiler.dirWrapperClassName(dirMeta.type);
-    const classStmt = new o.ClassStmt(wrapperClassName, null, fields, [], ctor, methods);
-    return new DirectiveWrapperCompileResult([classStmt], wrapperClassName);
+    return createClassStmt({
+      name: DirectiveWrapperCompiler.dirWrapperClassName(this.dirMeta.type),
+      ctorParams: dirDepParamNames.map((paramName) => new o.FnParam(paramName, o.DYNAMIC_TYPE)),
+      builders: [{fields: fields, ctorStmts: ctorStmts}, this]
+    })
```
Comment (vicb) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: ;

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts [hunk 12]
```diff
@@ -45,74 +46,83 @@ export class DirectiveWrapperCompiler {
...
-        null, dirDepParamNames.map((paramName) => new o.FnParam(paramName, o.DYNAMIC_TYPE)),
-        ctorStmts);
 
-    const wrapperClassName = DirectiveWrapperCompiler.dirWrapperClassName(dirMeta.type);
-    const classStmt = new o.ClassStmt(wrapperClassName, null, fields, [], ctor, methods);
-    return new DirectiveWrapperCompileResult([classStmt], wrapperClassName);
+    return createClassStmt({
+      name: DirectiveWrapperCompiler.dirWrapperClassName(this.dirMeta.type),
+      ctorParams: dirDepParamNames.map((paramName) => new o.FnParam(paramName, o.DYNAMIC_TYPE)),
+      builders: [{fields: fields, ctorStmts: ctorStmts}, this]
```
Comment (vicb) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: `{fields, ctorStmts}`

File: modules/@angular/compiler/src/ng_module_compiler.ts [hunk 13]
```diff
@@ -82,13 +83,15 @@ export class NgModuleCompiler {
   }
 }
 
-class _InjectorBuilder {
+class _InjectorBuilder implements ClassBuilder {
   private _tokens: CompileTokenMetadata[] = [];
```
Comment (vicb) [inline modules/@angular/compiler/src/ng_module_compiler.ts]: public then private

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts [hunk 14]
```diff
@@ -45,74 +46,83 @@ export class DirectiveWrapperCompiler {
...
+      createCheckInputMethod(inputFieldName, builder);
+    });
+    createDetectChangesInternalMethod(builder);
+    const classStmt = builder.build();
+    return new DirectiveWrapperCompileResult([classStmt], classStmt.name);
+  }
+}
+
+class DirectiveWrapperBuilder implements ClassBuilder {
+  fields: o.ClassField[] = [];
```
Comment (vicb) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: fields -> properties (TS terminology)

**Comment by tbosch** [comment]
Rename `FastArray` into `InlineArray`

File: modules/@angular/compiler/src/template_parser/binding_parser.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: Removed

File: modules/@angular/compiler/src/template_parser/binding_parser.ts [hunk 15]
```diff
@@ -0,0 +1,435 @@
...
+        const expression = hostProps[propName];
+        if (typeof expression === 'string') {
+          this.parsePropertyBinding(propName, expression, true, sourceSpan, [], boundProps);
+        } else {
+          this.reportError(
+              `Value of the host property binding "${propName}" needs to be a string representing an expression but got "${expression}" (${typeof expression})`,
+              sourceSpan);
+        }
+      });
+      boundProps.forEach(
```
Comment (tbosch) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: Keeping it for now

File: modules/@angular/compiler/src/template_parser/binding_parser.ts (hunk 2 above)
Comment (tbosch) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: Changed to return the array.

File: modules/@angular/compiler/src/template_parser/binding_parser.ts (hunk 3 above)
Comment (tbosch) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: Changed to return the array.

File: modules/@angular/compiler/src/template_parser/binding_parser.ts (hunk 4 above)
Comment (tbosch) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: The `_reportExpressionParserErrors` has an internal loop as well..

File: modules/@angular/compiler/src/template_parser/binding_parser.ts (hunk 5 above)
Comment (tbosch) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: changed

File: modules/@angular/compiler/src/template_parser/binding_parser.ts (hunk 6 above)
Comment (tbosch) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: I keep it for now, grouping by context.

File: modules/@angular/compiler/src/template_parser/binding_parser.ts [hunk 16]
```diff
@@ -0,0 +1,435 @@
...
+    const report = isAttr ? this._schemaRegistry.validateAttribute(propName) :
+                            this._schemaRegistry.validateProperty(propName);
+    if (report.error) {
+      this.reportError(report.msg, sourceSpan, ParseErrorLevel.FATAL);
+    }
+  }
+}
+
+export class PipeCollector extends RecursiveAstVisitor {
+  pipes: Set<string> = new Set<string>();
```
Comment (tbosch) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: changed

File: modules/@angular/compiler/src/template_parser/template_parser.ts (hunk 7 above)
Comment (tbosch) [inline modules/@angular/compiler/src/template_parser/template_parser.ts]: Done

File: modules/@angular/core/src/linker/view_utils.ts [hunk 17]
```diff
@@ -377,3 +377,153 @@ const CAMEL_CASE_REGEXP = /([A-Z])/g;
...
+  const el = renderer.createElement(parentElement, name, debugInfo);
+  for (var i = 0; i < attrs.length; i += 2) {
+    renderer.setElementAttribute(el, attrs.get(i), attrs.get(i + 1));
+  }
+  return el;
+}
+
+export function selectOrCreateRenderHostElement(
+    renderer: Renderer, elementName: string, attrs: FastArray<string>,
+    rootSelectorOrNode: string | any, debugInfo: RenderDebugInfo): any {
```
Comment (tbosch) [inline modules/@angular/core/src/linker/view_utils.ts]: changed

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts (hunk 9 above)
Comment (tbosch) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Keeping `build` for now

File: modules/@angular/compiler/src/compiler_util/identifier_util.ts [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+  } else {
+    return o.importExpr(token.identifier);
+  }
+}
+
+export function createFastArray(values: o.Expression[]): o.Expression {
+  if (values.length === 0) {
+    return o.importExpr(resolveIdentifier(Identifiers.EMPTY_FAST_ARRAY));
+  }
+  const index = Math.ceil(values.length / 2) - 1;
```
Comment (tbosch) [inline modules/@angular/compiler/src/compiler_util/identifier_util.ts]: Fixed and added tests.

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts [hunk 19]
```diff
@@ -45,74 +46,83 @@ export class DirectiveWrapperCompiler {
...
+  getters: o.ClassGetter[] = [];
+  methods: o.ClassMethod[] = [];
+  ctorStmts: o.Statement[] = [];
+
+  genChanges: boolean;
+  ngOnChanges: boolean;
```
Comment (tbosch) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Keeping the lifecycle hook method to make similar to the other properties.

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts (hunk 11 above)
Comment (tbosch) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Added

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts (hunk 12 above)
Comment (tbosch) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: changed

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts [hunk 20]
```diff
@@ -123,37 +133,43 @@ function detectChangesInternalMethod(
...
         new o.FnParam(RENDER_EL_VAR.name, o.DYNAMIC_TYPE),
         new o.FnParam(THROW_ON_CHANGE_VAR.name, o.BOOL_TYPE),
       ],
-      stmts, o.BOOL_TYPE);
+      stmts, o.BOOL_TYPE));
 }
 
-function checkInputMethod(
-    input: string, fieldExpr: o.ReadPropExpr, lifecycleHooks: GenConfig): o.ClassMethod {
+function createCheckInputMethod(input: string, builder: DirectiveWrapperBuilder) {
```
Comment (tbosch) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: changed

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts (hunk 14 above)
Comment (tbosch) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Talked offline, keeping fields, as we have o.ClassField and o.ClassGetter.

**Comment by IgorMinar** [comment]
this failed on TAP. @tbosch has the sponge link

**Comment by tbosch** [comment]
@IgorMinar I looked at the error and added a commit to fix it. Also tried it in G3 and it works.

**Comment by vsavkin** [comment]
Merged https://github.com/angular/angular/commit/178fb79b5cafe36c993bb1f9f9a63e8bac162c66

**Pull request closed without merging**